### ISSUE TITLE
[no-unused-vars] issue with varsIgnorePattern option

### ISSUE THREAD

**Issue by iotheo** [typescript-eslint/typescript-eslint#3649]
Using the `.` (dot) character in the `varsIgnorePattern` option, seems to match `any character` instead of `dot` itself.

This behavior is not consistent with the implementation of the [no-unused-vars](https://eslint.org/docs/rules/no-unused-vars), in which the `dot` matches the dot.

_I also tried to escape the dot character, but didn't seem to find any solution. If there's a way to do that, please let me know._

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
```JSON
{
  "rules": {
    "no-unused-vars": [
        1,
        {
          "ignoreRestSiblings": true,
          "argsIgnorePattern": "^_|...",
          "varsIgnorePattern": "^_|..."
        }
      ],
  }
}
```

```TS
import {a, b, ccc } from "./myModule";

const { d, e, ...rest } = justAnotherModule;
```

**Expected Result**

I expected to ignore the `...rest` variable.

**Actual Result**

It didn't ignore the `...rest`. It ignored the `ccc` as the `.` seems to match against _any_ character, which is not consistent with:

- [no-unused-vars](https://eslint.org/docs/rules/no-unused-vars)
-  the `argsIgnorePattern` option, which indeed, given a `.`, matches against a `.`. 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.3` |
| `@typescript-eslint/parser`        | `4.28.3` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.23.0` |
| `node`                             | `X.Y.Z` |

**Comment by bradzacher** [comment]
There's a lot to talk about with your post, as you're doing a lot of things incorrectly.

First of all - this is correct as it's how regexes work. The options are called ***pattern*** as they use regex patterns to match.
Dot means any character in a regex.

Second of all - you're misunderstanding variable naming. The variable is not called `...rest`, it's just called `rest`.
`...` is not part of the name - it is the rest operator.

Finally - you should not ignore rest variables either because unused rest variables are useless variables.

```ts
const {x, y, ...z} = someObject;
```

Is the same as

```ts
const {x, y} = someObject;
```

**Comment by iotheo** [comment]
Thank you for the input.

As mentioned in the initial post, the `argsIgnorePattern` matches against the `.` character but `varsIgnorePattern` doesn't, which is inconsistent.

I guess the difference in between is because var assignment cannot contain `.` 

Please verify that.

Have a nice day :)

**Comment by bradzacher** [comment]
Both an argument and a variable name cannot contain a `.`
It is an illegal character in an identifier.

But in a regex `.` can match a `.` because the token can match any character.

**Comment by iotheo** [comment]
Thanks for the answer.

> Second of all - you're misunderstanding variable naming. The variable is not called ...rest, it's just called rest.
... is not part of the name - it is the rest operator.

What confused me was what the pattern matches against, rather than was the variable name is. 

Makes total sense now.
